cpu/nrf52/periph/i2c.c

```diff
@@ -37,183 +41,176 @@
         }
-        if (dev(bus)->ERRORSRC & TWIM_ERRORSRC_DNACK_Msk) {
-            dev(bus)->ERRORSRC = TWIM_ERRORSRC_DNACK_Msk;
+        if (bus(dev)->ERRORSRC & TWIM_ERRORSRC_DNACK_Msk) {
+            bus(dev)->ERRORSRC = TWIM_ERRORSRC_DNACK_Msk;
             DEBUG("[i2c] check_error: NACK on data byte\n");
-            return -1;
+            return -EIO;
         }
     }
 
-    return len;
+    return 0;
 }
 
-int i2c_init_master(i2c_t bus, i2c_speed_t speed)
+void i2c_init(i2c_t dev)
 {
-    if (bus >= I2C_NUMOF) {
-        return -1;
-    }
-    if (speed & INVALID_SPEED_MASK) {
-        return -2;
-    }
+    assert(dev < I2C_NUMOF);
 
+    /* Initialize mutex */
+    mutex_init(&locks[dev]);
     /* disable device during initialization, will be enabled when acquire is
      * called */
-    dev(bus)->ENABLE = TWIM_ENABLE_ENABLE_Disabled;
-
+    bus(dev)->ENABLE = TWIM_ENABLE_ENABLE_Disabled;
     /* configure pins */
-    gpio_init(i2c_config[bus].scl, GPIO_IN_PU);
-    gpio_init(i2c_config[bus].sda, GPIO_IN_PU);
-    dev(bus)->PSEL.SCL = i2c_config[bus].scl;
-    dev(bus)->PSEL.SDA = i2c_config[bus].sda;
-
-    /* configure bus clock speed */
-    dev(bus)->FREQUENCY = speed;
+    gpio_init(i2c_config[dev].scl, GPIO_IN_PU);
+    gpio_init(i2c_config[dev].sda, GPIO_IN_PU);
+    bus(dev)->PSEL.SCL = i2c_config[dev].scl;
+    bus(dev)->PSEL.SDA = i2c_config[dev].sda;
+    /* configure dev clock speed */
+    bus(dev)->FREQUENCY = i2c_config[dev].speed;
 
     /* re-enable the device. We expect that the device was being acquired before
      * the i2c_init_master() function is called, so it should be enabled when
      * exiting this function. */
-    dev(bus)->ENABLE = TWIM_ENABLE_ENABLE_Enabled;
-
-    return 0;
+    bus(dev)->ENABLE = TWIM_ENABLE_ENABLE_Enabled;
+    return;
 }
 
-int i2c_acquire(i2c_t bus)
+int i2c_acquire(i2c_t dev)
 {
-    assert(bus < I2C_NUMOF);
+    assert(dev < I2C_NUMOF);
 
-    mutex_lock(&locks[bus]);
-    dev(bus)->ENABLE = TWIM_ENABLE_ENABLE_Enabled;
+    mutex_lock(&locks[dev]);
+    bus(dev)->ENABLE = TWIM_ENABLE_ENABLE_Enabled;
 
-    DEBUG("[i2c] acquired bus %i\n", (int)bus);
+    DEBUG("[i2c] acquired dev %i\n", (int)dev);
     return 0;
 }
 
-int i2c_release(i2c_t bus)
+int i2c_release(i2c_t dev)
 {
-    assert(bus < I2C_NUMOF);
+    assert(dev < I2C_NUMOF);
 
-    dev(bus)->ENABLE = TWIM_ENABLE_ENABLE_Disabled;
-    mutex_unlock(&locks[bus]);
+    bus(dev)->ENABLE = TWIM_ENABLE_ENABLE_Disabled;
+    mutex_unlock(&locks[dev]);
 
-    DEBUG("[i2c] released bus %i\n", (int)bus);
+    DEBUG("[i2c] released dev %i\n", (int)dev);
     return 0;
 }
 
-int i2c_read_byte(i2c_t bus, uint8_t addr, void *data)
+int i2c_write_regs(i2c_t dev, uint16_t addr, uint16_t reg,
+                   const void *data, size_t len, uint8_t flags)
 {
-    return i2c_read_bytes(bus, addr, data, 1);
+    assert((dev < I2C_NUMOF) && data && (len > 0) && (len < 255));
+
+    if(flags & (I2C_NOSTART | I2C_REG16 | I2C_ADDR10)) {
```

Comment:
missing space after if and why `I2C_NOSTART` is not supported ?

Reply:
I2C_NOSTART is NOT supported on this platform because of the way the hardware is designed. Unfortunately, when you give your data pointer and the length to the dedicated register, you cannot change this data pointer without issuing a repeated start.
nRF52 TWI hardware is very limited.